### PR TITLE
Fix ONNX loading in issues opencv#17516, opencv#17531

### LINKED ISSUE #17516: dnn module problems: InnerProductLayer
##### System information (version)

- OpenCV => 4.3.0
- Operating System / Platform => Ubuntu 18.04
- Compiler => Qt5

##### Detailed description

When I imported an ONNX model, I found that it failed at modules/dnn/src/layers/fully_connected_layer.cpp Line 122
```python
CV_CheckEQ(numOutput, inputs[0][cAxis - 1], "");
```
I have checked it, and I think it may be not right.
For example,
```python
class Decoder(nn.Module):
    def __init__(self, inPlanes=512, sDim=512, attDim=512):
        super(Decoder, self).__init__()
        self.inPlanes = inPlanes
        self.sDim = sDim
        self.attDim = attDim

        self.sEmbed = nn.Linear(sDim, attDim)
        self.xEmbed = nn.Linear(inPlanes, attDim)
        self.wEmbed = nn.Linear(attDim, 1)

    def forward(self, x):
        x = x.reshape(-1, self.inPlanes)
        xProj = self.xEmbed(x)
        state = torch.zeros(1, self.sDim)
        sProj = self.sEmbed(state)
        sProj = sProj.expand(25, self.attDim)
        sumTanh = torch.tanh(sProj + xProj)
        sumTanh = sumTanh.reshape(-1, self.attDim)
        vProj = self.wEmbed(sumTanh)
        vProj = vProj.reshape(1, 25)
        alpha = F.softmax(vProj, dim=1)
        context = torch.matmul(alpha, x)  # HERE
        return context
```
Set the shape of the input x to (1, 25, 512), then at 'HERE'
```python
        context = torch.matmul(alpha, x)  # HERE
```
The shape of alpha is (1, 25), and the shape of x is (25, 125), and both of them are not the constBlobs in OpenCV.
When I import it in OpenCV, in fully_connected_layer.cpp Line 122, we can find that:
- numOutput is 512
- inputs[0][cAxis - 1] is 1

Obviously, numOutput != inputs[0][cAxis - 1], and it throw an exception.

So I deleted the CV_CheckEQ, and loaded it successfully.

##### Steps to reproduce

It is described above.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

### LINKED ISSUE #17531: dnn module problems: Expand
##### System information (version)
- OpenCV =>4.30.
- Operating System / Platform =>Ubuntu 18.04
- Compiler =>Qt5

##### Detailed description

There may be a bug in importing an 'Expand' layer.

It can be reproduced with the network in https://github.com/opencv/opencv/issues/17516#issue-636053290.
Set the 'state' as an input of 'forward':
```python
def forward(self, x, state):
```
then, delete:
```python
state = torch.zeros(1, self.sDim)
```
In this manner, it will be copied and concatenated.
I think there may be a bug, because the name of the 'concat' layer is the same with the last copy of input.
I think we should add
 ``` cpp
node_proto.set_output(0, layerParams.name);
```
after https://github.com/opencv/opencv/blob/55ca0fcc279bf5257b104854feadacf8082f9302/modules/dnn/src/onnx/onnx_importer.cpp#L1221

If not, it will throw an exception in the following steps, because the concat layer can not be found with a wrong name.

##### Steps to reproduce

It is described above.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by sl-sergei**
**Merge with extra:** opencv/opencv_extra#793

resolves #17516
resolves #17531
 
Also resolves untracked problem with constant input for Gemm layers

```
opencv_extra=testdata_for_17516_17531

force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.4.0:16.04
build_image:Custom Win=openvino-2020.2.0
build_image:Custom Mac=openvino-2020.3.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by asmorkalov** [comment]
@sl-sergei Please add test for the failed case.

**Comment by dkurt** [comment]
Please target this patch to 3.4 branch

**Comment by asmorkalov** [comment]
@sl-sergei  Friendly reminder about test.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -902,6 +902,19 @@ void ONNXImporter::populateNet(Net dstNet)
...
+
+                LayerParams constParams;
+                constParams.name = node_proto.input(0);
+                constParams.type = "Const";
+                constParams.blobs.push_back(inputBuf);
+
+                opencv_onnx::NodeProto proto;
+                proto.add_output(constParams.name);
+                addLayer(dstNet, constParams, proto, layer_id, outShapes);
+            }
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:917]: Please add a test which covers this code. Thanks!

**Comment by dkurt** [comment]
Please avoid merge commits. We'd like to recommend to do rebase instead.

seems like there are failed tests: https://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux/builds

**Comment by dkurt** [comment]
Please take a look at failed tests: https://pullrequest.opencv.org/buildbot/builders/precommit_custom_windows/builds/1317

File: modules/dnn/test/test_onnx_importer.cpp [hunk 2]
```diff
@@ -606,6 +606,16 @@ TEST_P(Test_ONNX_layers, Pad2d_Unfused)
...
 
+TEST_P(Test_ONNX_layers, LinearWithConstant)
+{
+    testONNXModels("lin_with_constant");
+}
+
+TEST_P(Test_ONNX_layers, MatmulWithTwoInputs)
+{
+    testONNXModels("matmul_with_two_inputs");
+}
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:617]: Please skip the tests for all the versions lower 2020.4: 
2020.3 - failed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_mac/builds/1053
2020.4 - passed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/4279

Comment (sl-sergei) [reply modules/dnn/test/test_onnx_importer.cpp:617]: > Please skip the tests for all the versions lower 2020.4:
> 2020.3 - failed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_mac/builds/1053
> 2020.4 - passed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/4279

Thank you, I tried to reproduce it locally

Comment (dkurt) [reply modules/dnn/test/test_onnx_importer.cpp:617]: It's OK just add skip exception (see `applyTestTag`)

**Comment by dkurt** [comment]
```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] Test_ONNX_layers.LinearWithConstant/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.MatmulWithTwoInputs/1, where GetParam() = OCV/OCL_FP16
```
https://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux/builds/25739

**Comment by sl-sergei** [comment]
> ```
> [  FAILED  ] 2 tests, listed below:
> [  FAILED  ] Test_ONNX_layers.LinearWithConstant/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.MatmulWithTwoInputs/1, where GetParam() = OCV/OCL_FP16
> ```
> 
> https://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux/builds/25739

I see, it looks like it fails sporadically on Intel GPU's with OpenCL. There were no changes since the last successful run

**Review by dkurt — APPROVED** [review]
:+1: Thank you!

**Pull request merged**